### PR TITLE
Use checked math with CryptoStream buffer calculations

### LINKED ISSUE #47136: Arithmetic overflow in CryptoStream with very large reads
I noticed an issue while eyeballing code in `CryptoStream`, where attempting to perform very large reads can result in an arithmetic overflow if the input block size is greater than the output block size:

https://github.com/dotnet/runtime/blob/3946369dd6e42cc59e09651ce3e9d09751d44131/src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs#L329-L334

Say in the case of `FromBase64Transform`, it will divide the length by 3, then multiple by 4.

Simple repo:

```csharp
[Fact]
public static void EnormousRead()
{
    const int MaxLength = 0x60000000;
    const int MaxRaw = 0x5FFFFFD3;

    byte[] buffer = new byte[MaxRaw];
    byte[] output = new byte[MaxLength];

    buffer.AsSpan().Fill((byte)'A');

    using var stream = new MemoryStream(buffer);
    using var transform = new FromBase64Transform();
    using var cs = new CryptoStream(stream, transform, CryptoStreamMode.Read);
    cs.Read(output, 0, output.Length);
}
```

Granted this is probably not a typical way to use a stream.

As it is, the repro above will fail with:

```plain
Error Message:
   System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values. (Parameter 'minimumLength')
  Stack Trace:
     at System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1.Rent(Int32 minimumLength) in /code/personal/dotnet/runtime/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:line 87
```

### PR DESCRIPTION
**Pull request by bartonjs**
In .NET Framework, .NET 5, and .NET Core the block scaling logic was used for
`new byte[inputCount / inputBlockSize * outputBlockSize]` (or swap in/out).
That expression throws an OverflowException when the computed size
overflows.

Earlier in .NET 6 we changed to use ArrayPool.Rent.  When ArrayPool.Rent gets a
negative (overflowed) number it instead throws an ArgumentException.

This change restores the OverflowException, so that we're not leaking an
ArgumentException back to the caller.  (While the caller can control the
behavior with the arguments they pass to CryptoStream.Read/Write, the
description and paramName don't map in an obvious manner)

Making the scenario work has virtue, but requires a larger test investment than
we can make at this time.

Fixes #47136

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs [hunk 1]
```diff
@@ -317,6 +318,74 @@ public static void PaddedAes_PartialRead_Success()
             }
         }
 
+        [Fact]
```
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs:321]: We should make this conditional on using a 64-bit process, given we're trying to allocate a 1.6GB array.

**Review by stephentoub — APPROVED** [review]
Other than changing the [Fact]s to [ConditionalFact]s, LGTM. Thanks.

File: src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs:321]: Won't a 32-bit process just hit the try/OOM/skip?  My feeling was that the architecture doesn't really guarantee success or failure, so just accept "couldn't do it" and move on.

Or do you perhaps think 32-bit might get unfortunately close and then OOM after that point?

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Security.Cryptography.Primitives/tests/CryptoStream.cs:321]: I'm more worried it succeeds and then causes other tests in the process to start OOM'ing.

**Pull request merged**